src/jit/codegenxarch.cpp

```diff
@@ -8842,7 +8966,80 @@ void CodeGen::genPutArgStk(GenTreePtr treeNode)
 #endif // !_TARGET_X86_
 }
 
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
+#if defined(FEATURE_PUT_STRUCT_ARG_STK)
+// genStoreRegToStackArg: Store a register value into the stack argument area
+//
+// Arguments:
+//    type   - the type of value to be stored
+//    reg    - the register containing the value
+//    offset - the offset from the base (see Assumptions below)
+//
+// Notes:
+//    A type of TYP_STRUCT instructs this method to store a 16-byte chunk
+//    at the given offset (i.e. not the full struct).
+//
+// Assumptions:
+//    The caller must set the context appropriately before calling this method:
+//    - On x64, m_stkArgVarNum must be set according to whether this is a regular or tail call.
+//    - On x86, the caller must set m_pushStkArg if this method should push the argument.
+//      Otherwise, the argument is stored at the given offset from sp.
+//
+// TODO: In the below code the load and store instructions are for 16 bytes, but the
+//          type is EA_8BYTE. The movdqa/u are 16 byte instructions, so it works, but
+//          this probably needs to be changed.
+//
+void CodeGen::genStoreRegToStackArg(var_types type, regNumber srcReg, unsigned offset)
+{
+    assert(srcReg != REG_NA);
+    instruction ins;
+    emitAttr    attr;
+    unsigned    size;
+    if (type == TYP_STRUCT)
+    {
+        ins = INS_movdqu;
+        // This should be changed!
+        attr = EA_8BYTE;
+        size = 16;
+    }
+    else
+    {
+#ifdef FEATURE_SIMD
+        if (varTypeIsSIMD(type))
+        {
+            assert(genIsValidFloatReg(srcReg));
+            ins = ins_Store(type);
+        }
+        else
+#endif // FEATURE_SIMD
+#ifdef _TARGET_X86_
+             if (type == TYP_LONG)
+        {
+            assert(genIsValidFloatReg(srcReg));
+            ins = INS_movq;
+        }
+        else
+#endif // _TARGET_X86_
+        {
+            assert((varTypeIsFloating(type) && genIsValidFloatReg(srcReg)) ||
+                   (varTypeIsIntegralOrI(type) && genIsValidIntReg(srcReg)));
+            ins  = ins_Store(type);
+        }
+        attr = emitTypeSize(type);
+        size = genTypeSize(type);
+    }
+#ifdef _TARGET_X86_
+    if (m_pushStkArg)
+    {
+        inst_RV_IV(INS_sub, REG_SPBASE, size, EA_PTRSIZE);
```

Comment:
Oh no! I previously had separate code for 16-byte, long, floating, and other ints. I merged the code to make it cleaner, but obviously eliminated the push case. Since we have a separate issue for that (#6684), I think I will plan to address that in a subsequent PR.

---
In reply to: [79509623](https://github.com/dotnet/coreclr/pull/7252#discussion_r79509623) [](ancestors = 79509623)